### PR TITLE
Fix an additional case for custom ops for applicatives

### PR DESCRIPTION
**Pull request by dsyme**
There is a blog post on the `let! .. and! ...` feature here: https://github.com/dsyme/fsharp-presentations/blob/master/design-notes/rethinking-applicatives.md

Subsequent discussion led to adding [this section](https://github.com/dsyme/fsharp-presentations/blob/master/design-notes/rethinking-applicatives.md#appendix-using-custom-operations-with-let--and-) and uncovered a case that the current implementation doesn't deal with properly - custom operators on CEs that define a "BindReturn" (i.e. applicatives) but do not define a "Bind" (i.e. not monadic)

See https://github.com/fsharp/fslang-design/issues/335#issuecomment-587517636

This fixes this problem and adds a test case.  I'll add a clarification to the RFC.

I had to move a chunk of code to make the fix.

### CODE REVIEW

File: src/fsharp/TypeChecker.fs [hunk 1]
```diff
@@ -8471,6 +8373,14 @@ and TcComputationExpression cenv env overallTy mWhole (interpExpr: Expr) builder
                 let hasBindReturnN = not (isNil (TryFindIntrinsicOrExtensionMethInfo ResultCollectionSettings.AtMostOneResult cenv env bindRange ad bindReturnNName builderTy))
                 if hasBindReturnN && Option.isSome (convertSimpleReturnToExpr varSpace innerComp) then 
                     let consumePat = SynPat.Tuple(false, pats, letPat.Range)
+
+                    // Add the variables to the query variable space, on demand
+                    let varSpace = 
+                        addVarsToVarSpace varSpace (fun _mCustomOp env -> 
+                                use _holder = TemporarilySuspendReportingTypecheckResultsToSink cenv.tcSink
+                                let _, _, vspecs, envinner, _ = TcMatchPattern cenv (NewInferenceType()) env tpenv (consumePat, None) 
+                                vspecs, envinner)
```
Comment (cartermp) [inline src/fsharp/TypeChecker.fs]: Is this the core of the change, since the issue was not having access to the variable space?

Comment (dsyme) [reply src/fsharp/TypeChecker.fs]: There were two issues.  One was this (it only applied to Bind2Reaturn etc.). The other was in `convertSimpleReturnToExpr`. This function basically detects a thing that can be turned into a BindReturn.  However it wasn't accepting uses of custom operators.

To explain: a custom operator 

    b { part1; op; part2 }

sort of becomes

    b { let! varspace = op (b { part1; return varspace })
        part2 }

The `part1; return varspace` is eligible to become a `BindReturn`.  Previously we were requiring it to be `Bind` then `Return`.

**Comment by dsyme** [comment]
Can we approve and merge this please? Thanks

**Comment by cartermp** [comment]
The overall feature is protected

**Comment by dsyme** [comment]
> The overall feature is protected

That's correct

**Pull request merged**